Nos/Service/DatabaseCleaner.swift

```diff
@@ -122,5 +78,79 @@ enum DatabaseCleaner {
         let elapsedTime = Date.now.timeIntervalSince1970 - startTime.timeIntervalSince1970 
         Log.info("Finished Core Data cleanup in \(elapsedTime) seconds.")
     }
-    // swiftlint:enable function_body_length    
+    
+    /// This converts old hydrated events back to stubs. We do this because EventReferences can form long chains
+    /// of events that we can't delete. By stubbing an event we can delete its eventReferences and also the
+    /// referencedEvents.
+    private static func stubReferencedOldEvents(
+        before deleteBefore: Date, 
+        user: Author, 
+        in context: NSManagedObjectContext
+    ) throws {
+        let request = NSFetchRequest<Event>(entityName: "Event")
+        request.sortDescriptors = [NSSortDescriptor(keyPath: \Event.receivedAt, ascending: true)]
+        let oldEventPredicate = NSPredicate(format: "(receivedAt < %@)", deleteBefore as CVarArg)
```

Comment:
Do you think this predicate and the one below could be shared with the Event class (oldEventClause) so that we have just one NSCompoundPredicate(andPredicateWithSubpredicates) that uses oldEvent and referencedEvents.count > 0 as subpredicates and can be shared in both places? (I'm thinking about a public function that builds this predicate) While I think it doesn't do any magic, the code would be just a bit more organized, but if you think this would be over-engineering this, let me know, the code looks good enough to me.

Reply:
Very clear. Thanks for giving it a shot.